server/src/server.ts

```diff
@@ -324,6 +330,44 @@ connection.onCompletion((completionInfo: CompletionParams): CompletionItem[] =>
           const searchResult = findStatement(decl.body, null, { pos, genv: decls });
           if (searchResult === null || searchResult.environment === null) break;
 
+          const doc = documents.get(completionInfo.textDocument.uri);
+          if (doc) {
+            const context = getCompletionContext(
+              doc.getText(),
+              doc.offsetAt(completionInfo.position));
+
+            if (context) {
+              switch (context.tag) {
+                case CompletionContextKind.StructAccess:
+                  try {
+                    // Type safety? :D 
+                    const type = <ast.Type>synthExpression(decls, searchResult.environment, null, <ast.Expression>context.expr);
+                    const structname = getStructId(decls, type);
```

Comment:
We should differentiate between -> and . here with the result, and make sure that the -> has a struct pointer and the . has the regular struct, so if people are doing things wrong our completion won't help them to continue to do wrong things.

Reply:
Done